Small changes for basis, knot vector, domain iterator

hverhelst opened the pull request:
NEW:
Paraview write for boxes represented by matrices 
Nestedness check of knot vector
Extra constructor for `gsTensorDomainBoundaryIterator` 

IMPROVED: 

FIXED:
ID of `gsTensorDomainIterator`

API:
Make `gsBasis<T>::evalAllDersSingle_into` with argument `gsMatrix<T> &` deprecated, and add `gsBasis<T>::evalAllDersSingle_into` with argument `std::vector<gsMatrix<T>>` in line with `evalAllDers_into`. Only `gsBSplineBasis::evalAllDersSingle_into` overloads, but throws a warning for a bug anyways. 


----

Please consider the following checklist before issuing a pull
request:
- [X] Have you added an explanation of what your changes do and why
  you'd like us to include them?
- [X] Have you documented any new codes using Doxygen comments?
- [X] Have you written new tests or examples for your changes?
-----

hverhelst committed:
Add evaluation of single basis function
Add paraview write for boxes represented by matrices
Add nestedness check of knot vector
Add extra constructor for `gsTensorDomainBoundaryIterator`
FIX ID of `gsTensorDomainIterator`

filiatra reviewed (changes requested):
There is the function:

gsBasisFun<T> function(index_t i) const

that can be used to get a handle on a single basis function.
This should do the same as "computesingle".

Can you tryto use that and if it works the same, remove computeSinge ?
it looks a bit of duplicated functionality.

hverhelst commented:
![](https://github.trello.services/images/mini-trello-icon.png) [PR712](https://trello.com/c/7dCbOzSw/3-pr712)

hverhelst committed:
undo computeSingle

hverhelst commented:
> There is the function:
> 
> gsBasisFun function(index_t i) const
> 
> that can be used to get a handle on a single basis function. This should do the same as "computesingle".
> 
> Can you tryto use that and if it works the same, remove computeSinge ? it looks a bit of duplicated functionality.

I have reverted the `computeSingle`. There is still the `evalAllDersDingle_into`, but that one provides an implementation of an existing function with that name using different input arguments.

hverhelst reviewed (commented):
I have reverted the computeSingle. There is still the evalAllDersDingle_into, but that one provides an implementation of an existing function with that name using different input arguments.

hverhelst committed:
make `evalAllDersSingle_into` returning `gsMatrix<T> &` deprecated,
better is to use the version with `std::vector<gsMatrix<T>>`.
The only use of `evalAllDersSingle_into` is inside `gsBSplineBasis`,
which throws a bug warning.

hverhelst committed:
Merge remote-tracking branch 'origin/stable' into single_eval_and_co

hverhelst commented:
I have checked this PR again, @filiatra could you have a look.
One improvement would be to replace the `gsWriteParaview` function for boxes to a version where they are not written in `nBoxes` number of files, but instead just in one file (with an unstructured grid or so. I don't know yet which type to use in Paraview.

hverhelst committed:
remove `gsBasis<T>::evalAllDers_into`

hverhelst committed:
fix gsMapped(Single)Basis

hverhelst committed:
Update gsBasis.hpp
